bigtable/internal/transport/connpool.go

```diff
@@ -16,53 +16,600 @@ package internal
+	_, err := client.PingAndWarm(primeCtx, req, grpc.Peer(&p))
+	if err != nil {
+		return false, err
+	}
+
+	isALTS := false
+	if p.AuthInfo != nil {
+		if _, ok := p.AuthInfo.(alts.AuthInfo); ok {
+			isALTS = true
+		}
+	}
+	return isALTS, nil
+}
+
+func NewBigtableConn(conn *grpc.ClientConn, instanceName, appProfileId string) *BigtableConn {
+	return &BigtableConn{
+		ClientConn:   conn,
+		instanceName: instanceName,
+		appProfile:   appProfileId,
+	}
+}
+
+// probeResult stores a single health check outcome.
+type probeResult struct {
+	t          time.Time
+	successful bool
+}
+
+// connHealthState holds the health monitoring state for a single connection.
+type connHealthState struct {
+	mu               sync.Mutex // Guards fields below
+	probeHistory     []probeResult
+	successfulProbes int
+	failedProbes     int
+	lastProbeTime    time.Time
+}
+
+// addProbeResult records a new probe outcome and prunes old history.
+func (chs *connHealthState) addProbeResult(successful bool, windowDuration time.Duration) {
+	chs.mu.Lock()
+	defer chs.mu.Unlock()
+
+	now := time.Now()
+	chs.lastProbeTime = now
+	chs.probeHistory = append(chs.probeHistory, probeResult{t: now, successful: successful})
+
+	if successful {
+		chs.successfulProbes++
+	} else {
+		chs.failedProbes++
+	}
+	chs.pruneHistoryLocked(windowDuration)
+}
+
+// pruneHistoryLocked removes probe results older than WindowDuration. Assumes chs.mu is held.
+func (chs *connHealthState) pruneHistoryLocked(windowDuration time.Duration) {
+	windowStart := time.Now().Add(-windowDuration)
+	firstValid := 0
+	for firstValid < len(chs.probeHistory) && chs.probeHistory[firstValid].t.Before(windowStart) {
+		result := chs.probeHistory[firstValid]
+		if result.successful {
+			chs.successfulProbes--
+		} else {
+			chs.failedProbes--
+		}
+		firstValid++
+	}
+	if firstValid > 0 {
+		chs.probeHistory = chs.probeHistory[firstValid:]
+	}
+}
+
+// isHealthy reports if the connection is currently considered healthy based on probe history.
+func (chs *connHealthState) isHealthy(minProbesForEval int, failurePercentThresh int) bool {
+	chs.mu.Lock()
+	defer chs.mu.Unlock()
+	totalProbes := chs.successfulProbes + chs.failedProbes
+	if totalProbes < minProbesForEval {
+		return true // Not enough data to make a judgment
+	}
+	failureRate := float64(chs.failedProbes) / float64(totalProbes) * 100.0
+	return failureRate < float64(failurePercentThresh)
+}
+
+// getFailedProbes returns the number of failed probes in the current window.
+func (chs *connHealthState) getFailedProbes() int {
+	chs.mu.Lock()
+	defer chs.mu.Unlock()
+	return chs.failedProbes
+}
+
+// connEntry represents a single connection in the pool.
+type connEntry struct {
+	conn          *BigtableConn
+	unaryLoad     int32           // In-flight unary requests
+	streamingLoad int32           // Active streams
+	health        connHealthState // Embedded health state
+	altsUsed      int32           // Set to 1 atomically if ALTS is used, 0 otherwise.
```

Comment:
This thing is per-RPC, not per ClientConn. If the intention is to only save what's first PingAndWarm went through, this may be fine, but it gives a false impression that this conn uses ALTS for all calls.